include/ppx/geometry.h

```diff
@@ -157,43 +167,68 @@ class Geometry
         char*       GetData() { return DataPtr(mData); }
         const char* GetData() const { return DataPtr(mData); }
         uint32_t    GetElementCount() const;
+        uint32_t    GetDataSize() const;
 
         // Trusts that calling code is well behaved :)
         //
+        // Write(count, pValues...) is used to write a chunk of data at specified offset
+        // If called with offset=-1, will write at the end and adjust mOffset accordingly
         template <typename T>
-        void Append(const T& value)
+        void Write(uint32_t count, const T* pValues, int32_t offset = -1)
         {
-            uint32_t sizeOfValue = static_cast<uint32_t>(sizeof(T));
+            uint32_t sizeOfValues = count * static_cast<uint32_t>(sizeof(T));
+            uint32_t writeOffset  = offset == -1 ? mOffset : offset;
+
+            if (offset == -1 && !mFinalSizeSet) {
+                // New data is being appended, need to allocate storage
+                SetSize(writeOffset + sizeOfValues);
+            }
+            else {
+                PPX_ASSERT_MSG(writeOffset + sizeOfValues <= GetSize(), "Attempting to write values of size " << sizeOfValues << " at offset " << offset << " will overflow the buffer of size " << GetSize());
+            }
 
-            // Current size
-            size_t offset = mData.size();
-            // Allocate storage for incoming data
-            mData.resize(offset + sizeOfValue);
             // Copy data
-            const void* pSrc = &value;
-            void*       pDst = mData.data() + offset;
-            memcpy(pDst, pSrc, sizeOfValue);
+            const void* pSrc = pValues;
+            void*       pDst = mData.data() + writeOffset;
+            memcpy(pDst, pSrc, sizeOfValues);
+
+            if (offset == -1) {
```

Comment:
It looks like mOffset is not updated if offset is provided, what happens if offset > mOffset, does the mOffset become meaningless?